### PR TITLE
mon: cache encoded osdmaps for older clients

### PR DESCRIPTION
**Pull request by xiaoxichen**
By caching the backward-compatible encoded osdmap into LRUCache

### CODE REVIEW

**Comment by xiaoxichen** [comment]
more background about the initiative http://tracker.ceph.com/issues/23713

**Comment by tchaikov** [comment]
retest this please

**Comment by xiaoxichen** [comment]
@liewegas addressed, mind take another look?

File: src/mon/OSDMonitor.h [hunk 1]
```diff
@@ -43,6 +43,18 @@ class MOSDMap;
 #include "erasure-code/ErasureCodeInterface.h"
 #include "mon/MonOpRequest.h"
 
+namespace std
+{
+  template<>
+    struct hash<pair<version_t, uint64_t>> {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.h]: could just use `boost::hash`.

File: src/mon/OSDMonitor.h [hunk 2]
```diff
@@ -219,8 +231,8 @@ class OSDMonitor : public PaxosService {
 
   map<int,double> osd_weight;
 
-  SimpleLRU<version_t, bufferlist> inc_osd_cache;
-  SimpleLRU<version_t, bufferlist> full_osd_cache;
+  SimpleLRU<pair<version_t, uint64_t>, bufferlist> inc_osd_cache;
```
Comment (tchaikov) [inline src/mon/OSDMonitor.h]: could just use
```c++
using osdmap_key_t = std::pair<version_t, uint64_t>;
using osdmap_cache_t = SimpleLRU<osdmap_key_t,
				 bufferlist,
				 std::less<osdmap_key_t>,
				 boost::hash<osdmap_key_t>>;
osdmap_cache_t inc_osd_cache;
osdmap_cache_t full_osd_cache;
```

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -1300,7 +1300,7 @@ void OSDService::got_stop_ack()
 MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
                                                OSDSuperblock& sblock)
 {
-  MOSDMap *m = new MOSDMap(monc->get_fsid());
+  MOSDMap *m = new MOSDMap(monc->get_fsid(), osdmap->get_features(entity_name_t::TYPE_OSD, NULL));
```
Comment (tchaikov) [inline src/osd/OSD.cc]: wrap at 79 chars.

File: src/messages/MOSDMap.h [hunk 4]
```diff
@@ -63,14 +64,29 @@ class MOSDMap : public Message {
...
-
 public:
+  const static uint64_t significant_mask =
+           CEPH_FEATUREMASK_PGID64 |
+	   CEPH_FEATUREMASK_PGPOOL3 |
+	   CEPH_FEATUREMASK_OSDENC |
+           CEPH_FEATUREMASK_OSDMAP_ENC |
+	   CEPH_FEATUREMASK_MSG_ADDR2 |
+	   CEPH_FEATUREMASK_SERVER_LUMINOUS;
+  void set_features(uint64_t features) {
```
Comment (tchaikov) [inline src/messages/MOSDMap.h]: this method is not used.

File: src/messages/MOSDMap.h [hunk 5]
```diff
@@ -63,14 +64,29 @@ class MOSDMap : public Message {
...
+  MOSDMap(const uuid_d &f, const uint64_t features)
     : Message(CEPH_MSG_OSD_MAP, HEAD_VERSION, COMPAT_VERSION),
-      fsid(f),
+      fsid(f), features(features),
       oldest_map(0), newest_map(0) { }
 private:
   ~MOSDMap() override {}
-
 public:
+  const static uint64_t significant_mask =
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/messages/MOSDMap.h]: also, the name of `significant_mask` does not make sense. how about just name it `significant_features`?

File: src/messages/MOSDMap.h [hunk 6]
```diff
@@ -93,12 +109,7 @@ class MOSDMap : public Message {
     header.version = HEAD_VERSION;
     header.compat_version = COMPAT_VERSION;
     encode(fsid, payload);
-    if ((features & CEPH_FEATURE_PGID64) == 0 ||
-	(features & CEPH_FEATURE_PGPOOL3) == 0 ||
-	(features & CEPH_FEATURE_OSDENC) == 0 ||
-        (features & CEPH_FEATURE_OSDMAP_ENC) == 0 ||
-	(features & CEPH_FEATURE_MSG_ADDR2) == 0 ||
-	!HAVE_FEATURE(features, SERVER_LUMINOUS)) {
+    if (get_significant_features() == (features & significant_mask)) {
```
Comment (tchaikov) [inline src/messages/MOSDMap.h]: this is wrong. i'd recommend just put
```c++
if (features & significant_features == significant_features) {
 // ..
}
```

File: src/mon/OSDMonitor.cc [hunk 7]
```diff
@@ -1491,8 +1491,11 @@ void OSDMonitor::share_map_with_random_osd()
   }
 
   dout(10) << "committed, telling random " << s->inst << " all about it" << dendl;
+
+  //get features of the peer, fall back to my quorum_con_features.
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: add a space after `//`

Comment (tchaikov) [reply src/mon/OSDMonitor.cc]: and the wording of "fall back" is misleading, would be better to put

> use quorum_con_features, if it's an anonymous connection.

File: src/messages/MOSDMap.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (liewegas) [reply src/messages/MOSDMap.h]: Can you put this in OSDMap.h, and then use it for the mon's cache as well?

File: src/mon/OSDMonitor.cc [hunk 8]
```diff
@@ -3419,6 +3422,9 @@ void OSDMonitor::send_incremental(epoch_t first,
   dout(5) << "send_incremental [" << first << ".." << osdmap.get_epoch() << "]"
 	  << " to " << session->inst << dendl;
 
+  //get feature of the peer, fall back to my quorum_con_features.
+  uint64_t features= session->con_features ? session->con_features : mon->get_quorum_con_features();
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: stil nit, space between features and = and after //

File: src/mon/OSDMonitor.cc [hunk 9]
```diff
@@ -3496,12 +3496,53 @@ void OSDMonitor::get_removed_snaps_range(
...
+{
+  OSDMap m;
+  bufferlist::iterator q = bl.begin();
+  m.decode(q);
+  bl.clear();
+  m.encode(bl, features | CEPH_FEATURE_RESERVED);
+}
+
+int OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)
+{
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: features &= OSDMap::SIGNIFICANT_FEATURES;

File: src/mon/OSDMonitor.cc [hunk 10]
```diff
@@ -3496,12 +3496,53 @@ void OSDMonitor::get_removed_snaps_range(
...
+    if (inc_osd_cache.lookup(make_pair(ver, features), &bl)) {
       return 0;
     }
     int ret = PaxosService::get_version(ver, bl);
     if (!ret) {
-      inc_osd_cache.add(ver, bl);
+      if (features != mon->get_quorum_con_features()) {
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: & both feature sets with OSDMap::SIGNIFICANT_FEATURES before comparing

Comment (liewegas) [reply src/mon/OSDMonitor.cc]: or add and use a OSDMap::get_significant_features(features) helper?

File: src/mon/OSDMonitor.cc [hunk 11]
```diff
@@ -3624,20 +3665,27 @@ int OSDMonitor::get_full_from_pinned_map(version_t ver, bufferlist& bl)
...
-    }
-    if (ret != 0) {
-      return ret;
-    }
+  return get_version_full(ver, mon->get_quorum_con_features(), bl);
+}
 
-    full_osd_cache.add(ver, bl);
+int OSDMonitor::get_version_full(version_t ver, uint64_t features, bufferlist& bl)
+{
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: features &= OSDmap::SIGNIFICANT_FEATURES

File: src/mon/OSDMonitor.cc [hunk 12]
```diff
@@ -3624,20 +3665,27 @@ int OSDMonitor::get_full_from_pinned_map(version_t ver, bufferlist& bl)
...
+  }
+  int ret = PaxosService::get_version_full(ver, bl);
+  if (ret == -ENOENT) {
+    // build map?
+    ret = get_full_from_pinned_map(ver, bl);
+  }
+  if (ret != 0) {
+    return ret;
+  }
+  if (features != mon->get_quorum_con_features()) {
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: ...

**Comment by tchaikov** [comment]
@xiaoxichen you need following lines after `#include <boost/functional/hash.hpp>`

```
// re-include our assert to clobber the system one; fix dout:
#include "include/assert.h"
```
also, i don't think `boost/function.hpp` is necessary here.

like
```diff
diff --git a/src/mon/OSDMonitor.h b/src/mon/OSDMonitor.h
index 741771e..abc89fd 100644
--- a/src/mon/OSDMonitor.h
+++ b/src/mon/OSDMonitor.h
@@ -23,7 +23,10 @@

 #include <map>
 #include <set>
+#include <boost/functional/hash.hpp>

+// re-include our assert to clobber the system one; fix dout:
+#include "include/assert.h"
 #include "include/types.h"
 #include "include/encoding.h"
 #include "common/simple_cache.hpp"
@@ -42,20 +45,8 @@ class MOSDMap;

 #include "erasure-code/ErasureCodeInterface.h"
 #include "mon/MonOpRequest.h"
-//#include <boost/function.hpp>
-//#include <boost/functional/hash.hpp>
-
-namespace std
-{
-  template<>
-    struct hash<pair<version_t, uint64_t>> {
-      size_t operator()(const pair<version_t, uint64_t>& p) const {
-        size_t seed = hash<version_t>{}(p.first); {}
-        seed ^= hash<uint64_t>{}(p.second) + 0x9e3779b9 + (seed<<6) + (seed>>2);
-        return seed;
-      }
-    };
-};
+
+

 /// information about a particular peer's failure reports for one osd
 struct failure_reporter_t {
```

File: src/osd/OSDMap.h [hunk 13]
```diff
@@ -507,6 +507,14 @@ class OSDMap {
   int32_t max_osd;
   vector<uint32_t> osd_state;
 
+  static constexpr uint64_t SIGNIFICANT_FEATURES =
+            CEPH_FEATUREMASK_PGID64 |
+            CEPH_FEATUREMASK_PGPOOL3 |
+            CEPH_FEATUREMASK_OSDENC |
+            CEPH_FEATUREMASK_OSDMAP_ENC |
+            CEPH_FEATUREMASK_MSG_ADDR2 |
+            CEPH_FEATUREMASK_SERVER_LUMINOUS;
```
Comment (liewegas) [inline src/osd/OSDMap.h]: add SERVER_MIMIC here

File: src/osd/OSD.cc (hunk 3 above)
Comment (liewegas) [reply src/osd/OSD.cc]: this is the wrong method.  get_features() tells you what client featuers are required given what's in teh osdmap, not how the osdmap is/should be encoded.

I think we actaully needt o take this block from OSDMonitor::encode_pending() and move it into an OSDMap method, get_encode_features():
```
    // determine appropriate features
    if (tmp.require_osd_release < CEPH_RELEASE_LUMINOUS) {
      dout(10) << __func__ << " encoding without feature SERVER_LUMINOUS"
	       << dendl;
      features &= ~CEPH_FEATURE_SERVER_LUMINOUS;
    }
    if (tmp.require_osd_release < CEPH_RELEASE_KRAKEN) {
      dout(10) << __func__ << " encoding without feature SERVER_KRAKEN | "
	       << "MSG_ADDR2" << dendl;
      features &= ~(CEPH_FEATURE_SERVER_KRAKEN |
		    CEPH_FEATURE_MSG_ADDR2);
    }
    if (tmp.require_osd_release < CEPH_RELEASE_JEWEL) {
      dout(10) << __func__ << " encoding without feature SERVER_JEWEL" << dendl;
      features &= ~CEPH_FEATURE_SERVER_JEWEL;
    }
    if (tmp.require_osd_release < CEPH_RELEASE_MIMIC) {
      dout(10) << __func__ << " encoding without feature SERVER_MIMIC" << dendl;
      features &= ~(CEPH_FEATURE_SERVER_MIMIC);
    }
    dout(10) << __func__ << " encoding full map with " << features << dendl;
```
And the OSDMonitor code should just get an assert to make sure whatever the OSDMap method returns is a subset of the mon quorum features.

File: src/osd/OSD.cc [hunk 14]
```diff
@@ -1340,7 +1340,7 @@ void OSDService::send_incremental_map(epoch_t since, Connection *con,
     OSDSuperblock sblock(get_superblock());
     if (since < sblock.oldest_map) {
       // just send latest full map
-      MOSDMap *m = new MOSDMap(monc->get_fsid());
+      MOSDMap *m = new MOSDMap(monc->get_fsid(), osdmap->get_features(entity_name_t::TYPE_OSD, NULL));
```
Comment (liewegas) [inline src/osd/OSD.cc]: here too

File: src/messages/MOSDMap.h [hunk 15]
```diff
@@ -27,6 +27,7 @@ class MOSDMap : public Message {
 
  public:
   uuid_d fsid;
+  uint64_t encode_features;
```
Comment (tchaikov) [inline src/messages/MOSDMap.h]: please assign a default value to this member variable.

File: src/mon/OSDMonitor.cc [hunk 16]
```diff
@@ -2059,8 +2059,13 @@ bool OSDMonitor::preprocess_get_osdmap(MonOpRequestRef op)
 {
   op->mark_osdmon_event(__func__);
   MMonGetOSDMap *m = static_cast<MMonGetOSDMap*>(op->get_req());
+
+  uint64_t features = mon->get_quorum_con_features();
+  if (m->get_session() && m->get_session()->con_features)
+    features = m->get_session()->con_features;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: features = osdmap.get_encode_features()

File: src/mon/OSDMonitor.cc [hunk 17]
```diff
@@ -2059,8 +2059,13 @@ bool OSDMonitor::preprocess_get_osdmap(MonOpRequestRef op)
...
+
   dout(10) << __func__ << " " << *m << dendl;
-  MOSDMap *reply = new MOSDMap(mon->monmap->fsid);
+  MOSDMap *reply = new MOSDMap(mon->monmap->fsid, features);
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: I think this isn't quite right.. if you do this, you also need to pass features to the get_version calls below, right?

File: src/osd/OSD.cc (hunk 3 above)
Comment (xiaoxichen) [reply src/osd/OSD.cc]: but this is not a self contained piece, it based on 
```
  // features for osdmap and its incremental
  uint64_t features = mon->get_quorum_con_features();
```
so OSDMap itself cannot come out with a feature-set, but OSDMonitor find out proper features and use that to encode the map.

Comment (xiaoxichen) [reply src/osd/OSD.cc]: I think the issue is , if we get something from bl cache or from disk, do we have anyway to tell how the BL was encoded before fully decode it? 

the paxosservice::get_version in OSDMonitor looks like have same issue.

**Comment by liewegas** [comment]
see https://github.com/liewegas/ceph/commits/wip-osdmap-encode

File: src/osd/OSD.cc (hunk 3 above)
Comment (liewegas) [reply src/osd/OSD.cc]: not easily, no.  i added a comment in get_version[_full] to that effect. basically i think it doesn't matter.. we'll just (worst case) cache the same encoded osdmap under the latest mon quorum features and also the encoded osdmap's features.

**Comment by xiaoxichen** [comment]
@liewegas looks good,  yeah the worst case is we re-encoding once in monitor but several times(as no cache) in OSD, but it is fine at least OSD is much more parallelism than monitor.

c81e57beaf3d3c7d881db4c2ace8f263849bc6b9 fix the hashing

**Pull request closed without merging**